### PR TITLE
Build and push multiple (ARM/ARM64) architecture images on Docker Hub

### PR DESCRIPTION
**Pull request by CpuID**
This will make running Open Dynamic Export on smaller/embedded hardware easier, eg. Home Assistant Green/Blue/Yellow/Raspberry Pi + Apple Silicon

Originally I was going with `linux/arm64` only as an addition, but after [some reading](https://uninterrupted.tech/blog/creating-docker-images-that-can-run-on-different-platforms-including-raspberry-pi/) you also need specific arm/arm64 variants for some of the more powerful ARM devices out there. This may need extras added over time, but this change should cover the bulk of the required options.

https://en.wikipedia.org/wiki/Raspberry_Pi#Specifications shows most from a Pi2B upwards will work. Fairly sure the ODROID devices are `arm64/v8`, I'll confirm next week :)

Confirmed on MacOS it wants to use `linux/arm64/v8` with an Apple M1 Pro

### CODE REVIEW

**Comment by longzheng** [comment]
I don't have any experience with this but looks good. 

It's a bit hard fo rme to manually trigger the workflow on a remote branch so I'll just copy your branch first

**Comment by longzheng** [comment]
Test run https://github.com/longzheng/open-dynamic-export/actions/runs/12442729443

**Comment by CpuID** [comment]
The test pipeline takes a fair bit longer than the original one, there's no parallelisation in there right now.

If it's a problem there are ways to do it (separate steps/jobs per arch), but considering you only build/push for new tags, is it an issue currently...? If you're happy to just let it take the extra minutes in serial for now, that's fine :)

**Comment by longzheng** [comment]
I don't mind it taking a while, it's not a rush, however it seems to be stuck/frozen on this
```
#18 [linux/arm/v7 build 3/5] RUN --mount=type=bind,source=package.json,target=package.json     --mount=type=bind,source=package-lock.json,target=package-lock.json     --mount=type=cache,target=/root/.npm     npm ci
```
I cancelled the previous run after 15min https://github.com/longzheng/open-dynamic-export/actions/runs/12442729443/job/34741150224 just in case it was a once-off problem, but it seems to be stuck on the same platform in the new run https://github.com/longzheng/open-dynamic-export/actions/runs/12442729443/job/34741409521#step:6:402

**Comment by longzheng** [comment]
Looks like it may be this issue https://github.com/docker/build-push-action/issues/1071

However instead of downgrading the Node image I wonder if changing the GitHub Action runner image will also work

File: .github/workflows/publish-docker.yml [hunk 1]
```diff
@@ -43,6 +43,7 @@ jobs:
         with:
           context: .
           push: true
+          platforms: linux/amd64,linux/arm64,linux/arm/v7,linux/arm64/v8
```
Comment (longzheng) [inline .github/workflows/publish-docker.yml:46]: I saw here https://github.com/docker/build-push-action/blob/master/.github/build-push-summary.png that this also support a list format, which might be easier to understand

 ```suggestion
          platforms: 
            - linux/amd64
            - linux/arm64
            - linux/arm/v7
            - linux/arm64/v8
```

**Comment by longzheng** [comment]
Ah damn the macOS runner doesn't contain Docker https://github.com/actions/runner-images/blob/main/images/macos/macos-14-Readme.md

**Comment by CpuID** [comment]
Interesting... I wonder if we should skip `arm/v7` temporarily and get the other 2 in there at least? Then try figure out a solution for `arm/v7` as a separate PR?

**Comment by CpuID** [comment]
I agree downgrading Node is less than ideal...

**Comment by longzheng** [comment]
Yeah maybe we revisit `linux/arm/v7`, it sounds like there's some efforts/attempts to get this problem fixed upstream but I'm having a hard time following what exactly is the root cause and how far away the fix is.

**Comment by CpuID** [comment]
Yep seems reasonable 👍 did you want to kick off another test with https://github.com/longzheng/open-dynamic-export/pull/73/commits/6694b8d14805924fc6cb98f77579914ba776f561 ?

**Comment by CpuID** [comment]
Looking at https://en.wikipedia.org/wiki/Raspberry_Pi#Specifications it's only the Pi2B that's impacted by not having `arm/v7`

**Comment by longzheng** [comment]
> Yep seems reasonable 👍 did you want to kick off another test with [6694b8d](https://github.com/longzheng/open-dynamic-export/commit/6694b8d14805924fc6cb98f77579914ba776f561) ?

Hmmm now it's complaining about the new syntax https://github.com/longzheng/open-dynamic-export/actions/runs/12443121813/workflow

**Comment by CpuID** [comment]
Ah annoying lol, I'll adjust :)

**Comment by longzheng** [comment]
Oh wait the list syntax is not an array https://github.com/docker/build-push-action/tree/master?tab=readme-ov-file#inputs

```
List type is a newline-delimited string

cache-from: |
  user/app:cache
  type=local,src=path/to/dir
```

So I think it should be
```yaml
        with:
          context: .
          push: true
          # TODO: find a workaround for https://github.com/docker/build-push-action/issues/1071 without downgrading NodeJS
          # linux/arm/v7
          platforms: |
            linux/amd64
            linux/arm64
            linux/arm64/v8
```

**Comment by CpuID** [comment]
Most other projects out there tend to use the CSV syntax by the looks of it... https://github.com/search?q=%22platforms%3A%22+path%3A.github%2Fworkflows+language%3AYAML&type=code&l=YAML shall we just stick with it for consistency?

**Comment by CpuID** [comment]
Give https://github.com/longzheng/open-dynamic-export/pull/73/commits/a1c0dffb468e6cdc01529f8aa4affaec13308237 a try

**Comment by longzheng** [comment]
Sure sounds good, agree the list syntax is kind of unexpected

**Comment by CpuID** [comment]
I double checked, looks like Home Assistant Green and Blue are both `arm64/v8` hardware, and Apple Silicon is `arm64/v8` also. Home Assistant Yellow is a Pi4 which is the same again.

**Comment by longzheng** [comment]
It completed successfully but had a warning https://github.com/longzheng/open-dynamic-export/actions/runs/12443199747/job/34742129818#step:6:443

```
 1 warning found (use docker --debug to expand):
 - Duplicate platform result requested "linux/arm64"
```

**Comment by longzheng** [comment]
According to this `arm64/v8` is the same as `arm64` https://stackoverflow.com/a/70889505

**Comment by CpuID** [comment]
aha interesting, thx

**Comment by CpuID** [comment]
looks better https://hub.docker.com/r/longzhengau/open-dynamic-export/tags :)

**Comment by CpuID** [comment]
Can confirm it runs fine using [Orbstack](https://orbstack.dev/) on MacOS/Apple Silicon now 👍 handy for tuning configs locally

**Pull request merged**